src/execution/index/art/art.cpp

```diff
@@ -204,6 +204,160 @@ void ART::GenerateKeys(DataChunk &input, vector<unique_ptr<Key>> &keys) {
 	}
 }
 
+struct KeySection {
+	KeySection(idx_t start_p, idx_t end_p, idx_t depth_p, data_t key_byte_p)
+	    : start(start_p), end(end_p), depth(depth_p), key_byte(key_byte_p) {};
+	idx_t start;
+	idx_t end;
+	idx_t depth;
+	data_t key_byte;
+};
+
+void GetNodeChildren(vector<KeySection> &child_sections, vector<unique_ptr<Key>> &keys, KeySection &key_section) {
+
+	idx_t child_start_idx = key_section.start;
+	for (idx_t i = key_section.start + 1; i <= key_section.end; i++) {
+		if (keys[i - 1]->data[key_section.depth] != keys[i]->data[key_section.depth]) {
+			child_sections.emplace_back(child_start_idx, i - 1, key_section.depth + 1,
+			                            keys[i - 1]->data[key_section.depth]);
+			child_start_idx = i;
+		}
+	}
+	child_sections.emplace_back(child_start_idx, key_section.end, key_section.depth + 1,
+	                            keys[key_section.end]->data[key_section.depth]);
+}
+
+void Construct(vector<unique_ptr<Key>> &keys, row_t *row_ids, Node *&node, KeySection &key_section,
+               bool &has_constraint) {
+
+	auto &start_key = *keys[key_section.start];
+	auto &end_key = *keys[key_section.end];
+
+	// increment the depth until we reach a leaf or find a mismatching byte
+	auto prefix_start = key_section.depth;
+	while (start_key.len != key_section.depth && Key::ByteMatches(start_key, end_key, key_section.depth)) {
+		key_section.depth++;
+	}
+
+	// we reached a leaf, i.e. all the bytes of start_key and end_key match
+	if (start_key.len == key_section.depth) {
+
+		// end_idx is inclusive
+		auto num_row_ids = key_section.end - key_section.start + 1;
+
+		// check for possible constraint violation
+		if (has_constraint && num_row_ids != 1) {
+			throw ConstraintException("New data contains duplicates on indexed column(s)");
+		}
+
+		// new row ids of this leaf
+		auto new_row_ids = unique_ptr<row_t[]>(new row_t[num_row_ids]);
+		for (idx_t i = 0; i < num_row_ids; i++) {
+			new_row_ids[i] = row_ids[key_section.start + i];
+		}
+
+		Prefix prefix(start_key, prefix_start, start_key.len - prefix_start);
+		node = new Leaf(move(new_row_ids), num_row_ids, prefix);
+
+	} else { // create a new node and recurse
+
+		// we will find at least two child entries of this node, otherwise we'd have reached a leaf
+		vector<KeySection> child_sections;
+		GetNodeChildren(child_sections, keys, key_section);
+
+		auto node_type = Node::GetTypeBySize(child_sections.size());
+		Node::New(node_type, node);
+
+		auto prefix_length = key_section.depth - prefix_start;
+		node->prefix = Prefix(start_key, prefix_start, prefix_length);
+
+		// recurse on each child
+		for (idx_t i = 0; i < child_sections.size(); i++) {
```

Comment:
nitpick: Can this be a range-based for loop instead since we don't need `i` for anything other than `child_sections[i]` ?